apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc.rs

```diff
@@ -0,0 +1,254 @@
+        let used_space = self.protocol_txs.occupied_space_in_bytes
+            + self.encrypted_txs.occupied_space_in_bytes
+            + self.decrypted_txs.occupied_space_in_bytes;
+
+        self.block.occupied_space_in_bytes = used_space;
+
+        self.decrypted_txs = TxBin::default();
+        self.protocol_txs = TxBin::default();
+        self.encrypted_txs = TxBin::default();
+    }
+}
+
+/// Allotted space for a batch of transactions of the same kind in some
+/// proposed block, measured in bytes.
+#[derive(Debug, Copy, Clone, Default)]
+struct TxBin {
+    /// The current space utilized by the batch of transactions.
+    occupied_space_in_bytes: u64,
+    /// The maximum space the batch of transactions may occupy.
+    allotted_space_in_bytes: u64,
+}
+
+impl TxBin {
+    /// Return a new [`TxBin`] with a total allotted space equal to the
+    /// floor of the fraction `frac` of the available block space `max_bytes`.
+    #[inline]
+    fn init_over_ratio(max_bytes: u64, frac: Ratio<u64>) -> Self {
+        let allotted_space_in_bytes = (frac * max_bytes).to_integer();
+        Self {
+            allotted_space_in_bytes,
+            occupied_space_in_bytes: 0,
+        }
+    }
+
+    /// Return the amount of space left in this [`TxBin`].
+    #[inline]
+    fn space_left_in_bytes(&self) -> u64 {
+        self.allotted_space_in_bytes - self.occupied_space_in_bytes
+    }
+
+    /// Construct a new [`TxBin`], with a capacity of `max_bytes`.
+    #[inline]
+    fn init(max_bytes: u64) -> Self {
+        Self {
+            allotted_space_in_bytes: max_bytes,
+            occupied_space_in_bytes: 0,
+        }
+    }
+
+    /// Shrink the allotted space of this [`TxBin`] to whatever
+    /// space is currently being utilized.
+    #[inline]
+    fn shrink_to_fit(&mut self) {
+        self.allotted_space_in_bytes = self.occupied_space_in_bytes;
+    }
+
+    /// Try to dump a new transaction into this [`TxBin`].
+    ///
+    /// Signal the caller if the tx is larger than its max
+    /// allotted bin space.
+    fn try_dump<'tx>(&mut self, tx: &'tx [u8]) -> AllocStatus<'tx> {
+        let tx_len = tx.len() as u64;
+        if tx_len > self.allotted_space_in_bytes {
+            let bin_size = self.allotted_space_in_bytes;
+            return AllocStatus::OverflowsBin { tx, bin_size };
+        }
+        let occupied = self.occupied_space_in_bytes + tx_len;
+        if occupied <= self.allotted_space_in_bytes {
+            self.occupied_space_in_bytes = occupied;
+            AllocStatus::Accepted
+        } else {
+            let space_left = self.space_left_in_bytes();
+            AllocStatus::Rejected { tx, space_left }
+        }
+    }
+
+    /// Try to dump a new batch of transactions into this [`TxBin`].
+    ///
+    /// If an allocation fails, rollback the state of the [`TxBin`],
+    /// and return the respective status of the failure.
+    fn try_dump_all<'tx, T>(&mut self, txs: T) -> AllocStatus<'tx>
```

Comment:
Honestly I think it would be better not to try to share the same return type `AllocStatus` between this `try_dump_all` and `try_dump`, they are different enough. Returning `&tx` in `try_dump` isn't necessary, for `try_dump_all` it would be useful to know how far along the iteration we got (i.e. the "index" of the failing tx). I guess the caller can work this out given `&tx`, if they can reconstruct the iterator. I can suggest a code change here later maybe

Reply:
`try_dump_all` will be removed. it was only necessary to handle protocol txs, whose digests and vote extensions had to be allocated as a single unit. we're getting rid of digests, though.